frontend/packages/operator-lifecycle-manager/src/components/operator-hub/operator-hub-page.tsx

```diff
@@ -37,6 +49,107 @@ const ANNOTATIONS_WITH_JSON = [
+    (csv) => csv.metadata.name === `${obj.metadata.name}.${version}`,
+  );
+};
+
+const installPlanFor = (subscription, installPlans) => {
+  if (subscription) {
+    const installPlanName = subscription?.status?.installPlanRef?.name || '';
+    const installPlan: InstallPlanKind = installPlans?.data?.find(
+      (ip) => ip.metadata.name === installPlanName,
+    );
+    return installPlan;
+  }
+  return undefined;
+};
+
+const isOperatorInstalled = (
+  subscription,
+  operatorGroup,
+  packageName,
+  namespace,
+  clusterServiceVersion,
+  installPlan,
+) => {
+  const isInstalled = installedFor(subscription.data)(operatorGroup.data)(packageName)(namespace);
+  if (!isInstalled) {
+    return false;
+  }
+  // check for intermediate states
+  let loading = true;
+  let status = '';
+  let installObj: ClusterServiceVersionKind | InstallPlanKind = clusterServiceVersion;
+  const subscriptionData = subscription?.data;
+  status = installObj?.status?.phase;
+  if (installObj && status) {
+    loading = false;
+  } else if (subscriptionData) {
+    // There is no ClusterServiceVersion for the package, so look at Subscriptions/InstallPlans
+    loading = false;
+    status = subscriptionData?.status?.state || null;
+    const installPlanData: InstallPlanKind = installPlan;
+    if (installPlanData) {
+      installObj = installPlanData;
+    }
+  }
+  const isStatusSucceeded = status === 'Succeeded';
+  const isStatusFailed = status === 'Failed';
+  const isApprovalNeeded =
+    installObj?.spec?.approval === 'Manual' && installObj?.spec?.approved === false;
+
+  if (!loading && isStatusSucceeded) {
+    return true;
+  }
+  if (!loading && (isStatusFailed || isApprovalNeeded)) {
+    return false;
+  }
+
+  return isInstalled;
+};
+
+const isInstalling = (
+  subscription,
+  operatorGroup,
+  packageName,
+  namespace,
+  clusterServiceVersion,
+  installPlan,
+) => {
+  const isInstalled = installedFor(subscription.data)(operatorGroup.data)(packageName)(namespace);
+  if (!isInstalled) {
+    return false;
+  }
+  // check for intermediate states
+  let loading = true;
+  let status = '';
+  let installObj: ClusterServiceVersionKind | InstallPlanKind = clusterServiceVersion;
+  const subscriptionData = subscription?.data;
+  status = installObj?.status?.phase;
+  if (installObj && status) {
+    loading = false;
+  } else if (subscriptionData) {
+    // There is no ClusterServiceVersion for the package, so look at Subscriptions/InstallPlans
+    loading = false;
+    status = subscriptionData?.status?.state || null;
+    const installPlanData: InstallPlanKind = installPlan;
+    if (installPlanData) {
+      installObj = installPlanData;
+    }
+  }
+  const isStatusFailed = status === 'Failed';
+  const isApprovalNeeded =
+    installObj?.spec?.approval === 'Manual' && installObj?.spec?.approved === false;
+
+  if (!loading && (isStatusFailed || isApprovalNeeded)) {
+    return true;
+  }
+
+  return false;
+};
```

Comment:
See my main review comment about making the OperatorHubItem `installed` property into a string. Instead of representing multiple states with two booleans, we can just represent all of those states with a single string property.